### DISCUSSION TITLE
Add lazyQuery option

### DISCUSSION THREAD

**Discussion by krunaldodiya** [TanStack/query#393] [General] [answered]
Currently we have to pass args as second param and have to manually check if args is not undefined 

```javascript
const {data} = useQuery(['post', postId]);
```

it would be better to have full control on when to trigger query like for example searching posts

```javascript
const [searchPosts] = useLazyQuery('search_posts', searchPostsHandler);

searchPosts({
  variables: {
    post_id: 1
  }
})
```

**Comment by tannerlinsley** [comment]
Isn’t this the same as using the “manual“ option?

**Reply by krunaldodiya** [reply]
Yes, this was the main goal of the question to give more control on "how and when" to call useQuery

**Comment by tannerlinsley** [comment, marked as answer]
I see,  you don't want to pass the variables until you want to call it. With that expectation, how would `useLazyQuery` know what data to fetch from the cache? How would it know when to update?

I am having a hard time seeing how this pattern would benefit from the same things that useQuery and the like do. 

I'm going to turn this into a discussion so we can talk about the use cases more.

**Reply by krunaldodiya** [reply]
if i am not wrong, useQuery will automatically trigger when the props that is passed to it changes, this will make query too much depend on props, and handling this dynamically would be awkward too, for example:-

let say, i have list of country and foreach country i want to fetch cities of selected country

```javascript
const fetchCities = useLazyQuery('get_cities_by_country_id')

{ countries.map(country => (
     <div>
        <div>{country.name}</div>
        <button onClick={() => fetchCities({variables: {country_id: country.id}})}></button>
     </div>
   )
)}
```

How would you do the same fetching using useQuery() ?

**Reply by tannerlinsley** [reply]
Well, right now you wouldn't, but your example poses another question, what would you be gaining by using useQuery here instead of just a `usePromise`? Without reliable up-front keys, there can't be any cache hydration and what you're left with is simply a one-off promise manager no?

**Reply by cherniavskii** [reply]
@krunaldodiya regarding your example - how about this https://codesandbox.io/s/serene-fermat-86bdj?file=/src/App.js ?

**Reply by tannerlinsley** [reply]
Yes, this is probably the best way to do what you're talking about. This way you can use the cache, which is almost the entire point (and best feature IMO) of the library ;)

**Reply by krunaldodiya** [reply]
Got it,  seems like valid answer 👍
